descqa/NumberDensityVersusRedshift.py

```diff
@@ -0,0 +1,308 @@
+
+                if z0 is None and 'z0const' in self.validation_data.keys():
+                    z0 = self.validation_data['z0const'] + self.validation_data['z0linear'] * cut_lo
+            
+                z_this = catalog_data[self.z][mask]
+                del mask
+                total = '(# of galaxies = {})'.format(len(z_this))
+
+                #bin catalog_data
+                N,binEdges = np.histogram(z_this, bins=self.zbins)
+                sumz,binEdges = np.histogram(z_this, bins=self.zbins,weights=z_this)
+                meanz = sumz/N
+
+                #make subplot
+                ncol = int(n%self.ncolumns)
+                nrow = n//self.ncolumns
+                catalog_label = ' '.join([catalog_name, re.sub(self.band,filtername+' '+self.band,cutlabel)])
+                validation_label = ' '.join([self.validation_data.get('label', ''), cutlabel])
+                reskey = re.sub('\$','',cutlabel)
+                results[reskey] = self.make_subplot(meanz, z_this, n, ax_this, z0, z0err, catalog_color, validation_color, catalog_label, validation_label)
+                results[reskey]['total'] = total
+
+                #add curve for this catalog to summary plot
+                if self.first_pass:  #add validation data if evaluating first catalog
+                    summary = self.make_subplot(meanz, z_this, n, summary_ax_this, z0, z0err, catalog_color, validation_color, catalog_label, validation_label)
+                else:
+                    summary = self.make_subplot(meanz, z_this, n, summary_ax_this, 0., 0., catalog_color, validation_color, catalog_label, validation_label)
+            
+            else:
+                #make empty subplots invisible 
+                ax_this.set_visible(False)
+                summary_ax_this.set_visible(False)
+                
+        #save results for catalog and validation data in txt files
+        for filename, dtype, comment, info in zip_longest((filelabel, self.observation), ('y', 'fit'), (filtername,), ('total',)):
+            if len(filename)>0: 
+                with open(os.path.join(base_output_dir,'Nvsz_'+filename+'.txt'),'ab') as f_handle:     #open file in append mode
+                    #loop over magnitude cuts in results dict
+                    for key, value in results.items():
+                        self.save_quantities(dtype, value, f_handle, comment=' '.join(((comment or ''), key, value.get(info, ''))))
+
+        if self.first_pass: #turn off validation data plot in summary for remaining catalogs
+            self.first_pass = False
+
+        #save figure
+        plt.savefig(os.path.join(base_output_dir, 'Nvsz_'+filelabel+'.png'))
+        plt.close()
+        return TestResult(0, passed=True)
+
+    def make_subplot(self, meanz, catalog_data, nplot, f, z0, z0err, catalog_color, validation_color, catalog_label, validation_label):
```

Comment:
I feel like this function should be broken into 3 functions: `adjust_subplot_style`, `add_catalog_data_to_subplot`, and `add_validation_data_to_subplot`. Currently this `make_subplot` takes in lots of arguments, but not all of them are needed in each part. 

This way, the code in `run_on_single_catalog` would also become more readable. 

```python
self.adjust_subplot_style(summary_ax_this,  ...)
if self.first_pass:  #add validation data if evaluating first catalog
    self.add_validation_data_to_subplot(summary_ax_this, ...)
self.add_catalog_data_to_subplot(summary_ax_this, ...)
```

But I know this is quite a big change if you accept it, so I am ok deferring it to tomorrow.

Reply:
OK, I am not opposed to this, but let's discuss it tomorrow.